LibraryInstrumentationGuidance.md

```diff
@@ -0,0 +1,107 @@
+
+## Instrumentation 
+
+In the simplest case the operation that is being monitored needs to be wrapped by an activity. However in order to minimize performance impact the activity should be only created if there is any listener waiting for it
+
+```C#
+    static DiagnosticListener source = new DiagnosticListener("Example.MyLibrary");
+
+    Activity activity = null;
+    // create activity only when requested
+    if (source.IsEnabled("Example.MyLibrary.MyOperation"))
+        activity = source.StartActivity("Example.MyLibrary.MyOperation", new { Input = input });
+
+    try
+    {
+        // perform the actual operation
+        var output = RunOperation(input);  
+    }
+    finally
+    {
+        // stop activity if started
+        if (activity != null)
+             source.StopActivity(activity, new { Input = input, Output = output }); 
+    }
+```
+> ### *__TODO__ - provide a pointer to a document with more advanced instrumentation examples (WIP)* 
+
+### Payload
+
+When starting and stopping an activity it is possible to pass a payload object which will be available for the listener. This allows the listener to access raw, non-serialized data in real time. It can be useful for pulling out additional diagnostic information but also manipulating data as it is being processed (e.g. inject diagnostic context before an outbound call is made). Good examples of such payload are ```HttpRequestMessage``` or messages that are being passed through queues.
+
+Mind that payload is not preserved as part of Activity and is only available when activity is started/stopped so it is a good practice to specify all data that was passed to ```StartActivity()``` in ```StopActivity()``` as well.
+
+#### Payload format
+
+Diagnostic source event and activity start/stop API allows to specify only a single payload object, however in order to pass more data and allow future additions the recommendation is to use dynamic objects. Since these are .NET objects the names of particular properties should follow [standard .NET naming convention][DotNetPropertyNamingConvention]. 
+
+Here are some recommendations for typical payload property names:
+
+| Property name | Description |
+|:--------------|:-------------------|
+| `Endpoint` | The ```Uri``` of an endpoint the activity is for (e.g. target database, service) |
+| `PartitionKey` | The key/ID of the partition the activity is for |
+| `Status` | The ```TaskStatus``` of a completed asynchronous task |
+| `Exception` | The captured exception object |
+
+### Performance
+
+In order to avoid unnecessary overhead it is highly recommended to check if there is any listener for given activity. This can be done by calling 
```

Comment:
Please also mention that any instrumentation should start with DiagnosticListener.IsEnabled() - no args. It checks for ANY listener for this source and is very efficient. IsEnabled(string) is virtual method and does not give zero instrumentation cost.

Reply:
Good point - added that remark and updated code samples